### DISCUSSION TITLE
How can i create tables from my app dynamically?

### DISCUSSION THREAD

**Discussion by Ansh-Rathod** [supabase/supabase#3852] [Questions] [answered]
hi I'm creating a chat application with supabase in flutter and I need to create new chat rooms manually from the app and for that, I need to create a new table for each room. so how can I run a create table query from my app and create a table in the database? is it possible or please suggest any alternative solutions :)

**Comment by silentworks** [comment, marked as answer]
This isn't how you work with relational databases, you would create a table called rooms and add entries for each room in that, you would also create a table for messages which would then have a column called `room_id` as a relationship to the room. There is a slack chat example in the Supabase dashboard when you create a new project, take a look at that.

**Reply by LaughingBubba** [reply]
OK well me three. 

I dunno about dynamic, but lets says for example I want create a new supabase instance and create all the same tables I have from say a locally hosted PG docker container I created a few years ago. Are you saying the only way to create those would be to copy+paste my "CREATE TABLE ..." statements from the OG database into the the web based SQL Editor? What if I had literally a 1,000 tables?

OR can I execute those statements from a client? (no trigger function chicanery)

**Comment by ragunaru** [comment]
Sorry for necroing an old thread, but I do actually have this exact question.

In my specific use case I would need to have a custom table for each of my users where they control the schema. A high level example: each user has their own CRM table with unique columns.

Unless, of course, there's a smarter way of doing this, in which case I'd be very happy with pointers.

Thanks!

**Reply by Carlchk** [reply]
I think you can write a function for the Postgres database, the job of this function is to create a new table for your schema, and you can use rpc to call the create table function.

How to create a function in postgres db:
https://www.postgresql.org/docs/9.1/sql-createfunction.html

How rpc use in supabase with javascript:
https://supabase.com/docs/reference/javascript/rpc

**Reply by GaryAustin1** [reply]
The alternative is to store custom crm rows as json or jsonb in a table with id, user_id, crm_json columns.
Have each crm row encoded as json into the common table with the user_id as the main filter/RLS and rows for each crm row.

Supabase has operators for querying json https://supabase.com/docs/reference/javascript/select#querying-json-data
Postgres has extensive json operators including searching.  https://www.postgresql.org/docs/14/functions-json.html.

**Reply by vinliao** [reply]
Or, another alternative is... MongoDB. Different tool for different job.

**Reply by ragunaru** [reply]
> The alternative is to store custom crm rows as json or jsonb in a table with id, user_id, crm_json columns. Have each crm row encoded as json into the common table with the user_id as the main filter/RLS and rows for each crm row.
> 
> Supabase has operators for querying json https://supabase.com/docs/reference/javascript/select#querying-json-data Postgres has extensive json operators including searching. https://www.postgresql.org/docs/14/functions-json.html.

This idea actually led me to an alternative solution which I will test, so thanks!

**Comment by 0xc00010ff** [comment]
I also have this same need for a CMS kind of application where at least table-level multi-tenancy is a requirement. We would be talking to Supabase from a secured Node server. I figured this is probably a job for our platform-level cloud provider but I want to exhaust all possible options with Supabase. 

 @Carlchk thanks for the hint, we will explore that. 
The json idea is plausible, but in our case we would strongly prefer to keep consistent high-level table semantics for the data. 

Any other ideas for programmatically executing platform-level operations is appreciated. 

Edit: This API _exists_- it's the private API Supabase's platform calls. There's no public version of that right? Where I can `POST /columns` and other admin-level operations?

**Comment by WalnutIT** [comment]
Hello, 

you can achieve that by using a trigger function. I did following thing: 

1. I already had the trigger function, which was looking like this:
```plpgsql
-- inserts a row into public.users
create function public.handle_new_user() 
returns trigger 
language plpgsql 
security definer set search_path = public
as $$
begin
  insert into public.profiles(id, user_name)
  values(new.id, new.raw_user_meta_data->>'user_name');
  return new;
  
end;
$$;

-- trigger the function every time a user is created
create trigger on_auth_user_created
```

2. I created a table 'inbox' with the columns I need, via the supabase dashboard.

3. I extended the "handle_new_user" function with following code, which creates a user related inbox, when a user signs up: 

```plpgsql
-- create a user inbox
  _user_id_table_format = REPLACE(new.id::text,'-','_');

  EXECUTE format('CREATE TABLE public.%I () INHERITS (public.inbox)', 'inbox_' || _user_id_table_format);

  EXECUTE format('alter table public.%I enable row level security', 'inbox_' || _user_id_table_format);

  
  EXECUTE format('CREATE POLICY "Enable insert for authenticated users only" ON public.%I
  AS PERMISSIVE FOR INSERT
  TO authenticated

  WITH CHECK (true)', 'inbox_' || _user_id_table_format);

  EXECUTE format('CREATE POLICY "Enable read access for inbox owner" ON public.%I
  AS PERMISSIVE FOR SELECT
  TO authenticated
  USING (auth.uid() = inbox_owner)', 'inbox_' || _user_id_table_format);
  -- end creating inbox
```

The result is a new table with the name "inbox_<user_id>". I am using table partitioning: https://www.postgresql.org/docs/current/ddl-partitioning.html#:~:text=PostgreSQL%20allows%20you%20to%20declare,used%20as%20the%20partition%20key. 

I saw the example, which was mentioned by @silentworks, but I thought that partitioning is maybe the better and faster way to handle a large amount of data, but I don't know yet, if I will use this solution, because I want to use webhooks in order to use OneSignal to send push notifications to the users, when they receive a new message in their inbox. No idea, if that is possible to realise. I will see that in the next days.

**Comment by c4spar** [comment]
Hi, I am also looking for a way to create tables dynamically.

My use case would be to create all requried tables on the first launch of my application, so I don't need to create the tables manually via the ui or using a second database client.
The application should just work on first launch with no manual setup required.

**Reply by silentworks** [reply]
This is something you would use database migrations for.

**Reply by WalnutIT** [reply]
@silentworks I think he means sth. else. 

@c4spar I assume, that you need a database version control management like liquibase. 

The goal of this database version control management is, that it runs the sql statements, which you defined and keeping your database structure up to date. 

I am using a self designed json file to organize my sql statements, which looks like this: 

```JSON
{
  "databaseChangeLog": [
    {
      "changeSet": {
        "sequenceNumber": "1",
        "author": "Your name",
        "date": "16.12.2022",
        "changes": [
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS change_set (change_set_id TEXT PRIMARY KEY,sequence_number INTEGER NOT NULL UNIQUE,author TEXT,creation_date TEXT NOT NULL,execution_time TEXT NOT NULL)"
          },
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS statement (statement_id TEXT PRIMARY KEY,statement TEXT NOT NULL,change_set_id Text NOT NULL,FOREIGN KEY (change_set_id) REFERENCES change_set(change_set_id))"
          },
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS local_image_meta_data (local_image_meta_data_id TEXT PRIMARY KEY, path String NOT NULL UNIQUE, google_place_id TEXT NOT NULL, creation_date_time TEXT NOT NULL, is_public INTEGER NOT NULL, is_sight INTEGER NOT NULL, is_natural_feature INTEGER NOT NULL)"
          },
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS statistics (statistics_id TEXT PRIMARY KEY, google_places_id TEXT NOT NULL, city TEXT NOT NULL, country TEXT NOT NULL)"
          }
        ]
      }

    },
    {
      "changeSet": {
        "sequenceNumber": "2",
        "author": "Your name",
        "date": "09.03.2023",
        "changes": [
          {
            "sqlStatement": "ALTER TABLE local_image_meta_data ADD temperature REAL"
          },
          {
            "sqlStatement": "ALTER TABLE local_image_meta_data ADD unit TEXT"
          }
        ]
      }
    },
    {
      "changeSet": {
        "sequenceNumber": "3",
        "author": "Your name",
        "date": "12.04.2023",
        "changes": [
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS community_news_short_info (id TEXT PRIMARY KEY, created_at TEXT NOT NULL, type TEXT NOT NULL, image_url TEXT NOT NULL, local_image_path TEXT NOT NULL, en_title TEXT NOT NULL, en_subtitle TEXT NOT NULL, de_title TEXT NOT NULL, de_subtitle TEXT NOT NULL,content_key TEXT NOT NULL, is_new INTEGER NOT NULL)"
          }
        ]
      }
    },
    {
      "changeSet": {
        "sequenceNumber": "4",
        "author": "Your name",
        "date": "09.05.2023",
        "changes": [
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS messages (id TEXT PRIMARY KEY, created_at TEXT NOT NULL, inbox_owner TEXT NOT NULL, sender TEXT NOT NULL, message_content_id TEXT, message_type TEXT NOT NULL, read INTEGER NOT NULL)"
          }
        ]
      }
    },
    {
      "changeSet": {
        "sequenceNumber": "5",
        "author": "Your name",
        "date": "06.06.2023",
        "changes": [
          {
            "sqlStatement": "CREATE TABLE IF NOT EXISTS tp_user_card_cache_data (id TEXT PRIMARY KEY, favorite_location TEXT NOT NULL, travel_motto TEXT NOT NULL, user_name TEXT NOT NULL, avatar_image_path TEXT, user_id TEXT NOT NULL, avatar_name TEXT NOT NULL, is_suggested INTEGER NOT NULL, updated_at text NOT NULL)"
          }
        ]
      }
    }
  ]
}
```

It is an array of change sets for my database. The table "change_set" is mandatory. Here we store the meta data (especially the last executed sequence number). This allows us to detect with each start of the application, if a new sequence number was added. If yes, we execute them all. E. g. last stored sequence number was 3 and with the new release the sequences 4 and 5 were added, then the app will execute the statements of this sequence. I also use the table "statements", where I store the executed statements. I thought it can be useful for error detection. Until now I never used it, but I am keeping it for documentation reason. 

I am using that with my local SQFLite database (for mobile devices), but the same procedure also can be used with Supabase. E. g. you can use a RPC call and pass the statements, which have to be executed (well ordered of course) via an text array to the function and execute it there. Here are many options how you can do it. 

**Summary**
1. Add a database  change log file to your application (e. g. like the json above)
2. Check with each start, if new sequences were added
3. If new sequences were added, update the database and also the change log meta data. 

I hope it fits your needs. If not, let me know it.
